### PR TITLE
MitoHiFi module

### PR DESCRIPTION
**Pull request by verku**

### CODE REVIEW

**Comment by verku** [comment]
I've implemented MitoHiFi to assemble mitochondrial genomes from contigs, the alternative option would be to assemble them from the PacBio reads. I decided to do so because then it's possible to use the new MitoHiFi flag `-a plant` which allows for the assembly of a plant mitogenome. This feature is not optimized yet in the current MitoHiFi version though (see https://github.com/marcelauliano/MitoHiFi#new-parameter-for-plants).

File: modules/local/mitohifi/mitohifi.nf [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'docker://biocontainers/mitohifi:2.2_cv1' }"
+
+    input:
+    // TODO nf-core: Where applicable all sample-specific information e.g. "id", "single_end", "read_group"
+    //               MUST be provided as an input via a Groovy Map called "meta".
+    //               This information may not be required in some instances e.g. indexing reference genome files:
+    //               https://github.com/nf-core/modules/blob/master/modules/bwa/index/main.nf
+    tuple val(meta), path(mitoref)
+    tuple val(meta), path(gb)
```
[the remark this answers is not in the document]
Comment (mahesh-panchal) [reply modules/local/mitohifi/mitohifi.nf:29]: What is `mitoref`?

File: modules/local/mitohifi/findmitoref.nf [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+//               Any parameters that need to be evaluated in the context of a particular sample
+//               e.g. single-end/paired-end data MUST also be defined and evaluated appropriately.
+process FINDMITOREF {
+    tag "$meta.id"
+    label 'process_low'
+
+    // There is no conda package available for MitoHiFi
+    // Adding the Docker container from biocontainers found on DockerHub
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'docker://biocontainers/mitohifi:2.2_cv1' }"
```
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/findmitoref.nf:18]: Thanks! I was already wondering about the syntax here.

File: modules/local/mitohifi/findmitoref.nf [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    input:
+    // TODO nf-core: Where applicable all sample-specific information e.g. "id", "single_end", "read_group"
+    //               MUST be provided as an input via a Groovy Map called "meta".
+    //               This information may not be required in some instances e.g. indexing reference genome files:
+    //               https://github.com/nf-core/modules/blob/master/modules/bwa/index/main.nf
+    tuple val(meta)
```
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/findmitoref.nf:25]: The tool findMitoReference.py does not have any input besides the argument `--species`, and some optional arguments (email, outfolder, min_length, type of organelle). Should there be no input specified at all in that case?

File: modules/local/mitohifi/findmitoref.nf [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    // TODO nf-core: It MUST be possible to pass additional parameters to the tool as a command-line string via the "task.ext.args" directive
+    """
+    findMitoReference.py \\
+        $args \\
+        --outfolder /data/ \\
```
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/findmitoref.nf:43]: Thanks, this wasn't clear to me! I got `/data/` from the instructions how to run MitoHiFi (and findMitoReference) from a singularity image where they suggested this command: 

`singularity exec --bind /path/on/disk/to/data/:/data/ /path/to/mitohifi-v2.2.sif  findMitoReference.py --species "Cryptosula pallasiana" --email your@email.for.ncbi.db.query --outfolder /data/ --min_length 16000`

File: modules/local/mitohifi/mitohifi.nf (hunk 1 above)
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/mitohifi.nf:29]: Both are outputs from `findmitoref.nf`, so from the search for a mitochondrial reference genome for MitoHiFi. `gb` is the mitochondrial genome in genebank format, `mitoref` is the mitochondrial genome in fasta format (could be called fasta, too, but I wasn't sure at that point if something else will be called fasta and wanted to be more specific).

[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/mitohifi.nf:29]: > The second instance of meta here will cause issues as it's the same variable as the line above. If `gb` should be paired by the `meta` data with `mitoref` then it's done with a channel operation before being passed to the process, and then the resulting channel input changed to expect both.

I used a similar syntax for the output channel for findmitoref, does this also need to be changed?

File: modules/local/mitohifi/mitohifi.nf [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    // TODO:         Fix input channel if necessary. Input can be fasta file with raw reads (i.e. can be more than one file) 
+    //               or fasta file with contigs (one file). Implement assembly from contigs for now.
+    tuple val(meta), path(fasta)
+
+    output:
+    tuple val(meta), path("final_mitogenome.fasta"), emit: fasta
```
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/mitohifi.nf:35]: So you mean like that?

`tuple val(meta), path("{prefix}.fasta"), emit: fasta`

The output file names from MitoHiFi are named with default names like `final_mitogenome.*`, `contigs_stats.tsv` (and a number of output folders that I didn't add as output channels), I didn't find an argument in MitoHiFi to change the file names.

File: nextflow.config [hunk 6]
```diff
@@ -40,6 +40,20 @@ params {
...
+    email              = null
+    // minimum length of the mitochondrial genome 
+    // from NCBI that will be used as reference for MitoHiFi
+    min_length         = 16000
+    // organism genetic code following NCBI table
+    // see https://github.com/marcelauliano/MitoHiFi#parameter-list
+    o                  = null
+    // organism group. Values: 'animal' (default), 'plant' (not optimized yet)
+    a                  = null
+    // add more parameters from MitoHiFi?
```
[the remark this answers is not in the document]
Comment (verku) [reply nextflow.config:56]: I don't really understand how `ext.args` works, to be honest. I'll make a draft of this in the `modules.config` file. Do we leave these args as they are in the `nextflow.config`, for the user to be specified?

File: modules/local/mitohifi/mitohifi.nf [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    tuple val(meta), path("final_mitogenome.gb")   , emit: gb
+    tuple val(meta), path("contigs_stats.tsv")     , emit: tsv
```
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/mitohifi.nf:37]: MitoHiFi only provides a fasta of the mitogenome it found and circularized, so if we want to remove the mitogenome from the genome assembly we need to do that separately.

[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/mitohifi.nf:37]: There are more outputs than I specified in the output channel available (see everything here: https://github.com/marcelauliano/MitoHiFi#outputs). Should this also be in the output?

"Folder final_mitogenome_choice will contain a few files, the most important one is

- all_mitogenomes.rotated.aligned.fa - this is an aligment of all the mithocondrial sequences assembled by the pipeline. It is possible you will find heteroplasmy in your sample, in which case you will have more than one version of the final mito presented. The pipeline chooses a final representative by a majority rule, using cdhit-est to cluster sequenvces at a 80% identitty and chosing the largest one in that cluster as the final. If you want to study heteroplasmy of your sample, please investigate the all_mitogenomes.rotated.aligned.fa file further, and all the results in the potential_contigs folder."

File: modules/local/mitohifi/findmitoref.nf (hunk 4 above)
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/findmitoref.nf]: I just made a test run on the command line to see if we really need the flag `--outfolder` and findMitoReference.py requires it. To me it looks like `--outfolder` should point to the directory that is bound to the container. How do we solve this in Nextflow?

[the remark this answers is not in the document]
Comment (mahesh-panchal) [reply modules/local/mitohifi/findmitoref.nf]: Does the tool require the `--outfolder` to be `/data/` when running in singularity or can it have a different value?

File: modules/local/mitohifi/findmitoref.nf (hunk 3 above)
[the remark this answers is not in the document]
Comment (mahesh-panchal) [reply modules/local/mitohifi/findmitoref.nf:24]: Those inputs should then likely be `val` types that are passed in from channels. It's likely they'll be in the `meta` object but we should make explicit what the tool is expecting as input, and then use a channel operator to pass the correct `meta` field to this input channel.

File: nextflow.config (hunk 6 above)
[the remark this answers is not in the document]
Comment (verku) [reply nextflow.config]: I've made a draft in `modules.config` but I was unsure about the `prefix` (see my comments in the file)

File: modules/local/mitohifi/findmitoref.nf (hunk 3 above)
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/findmitoref.nf]: I've made a suggestion, see commit [fe7a689](https://github.com/NBISweden/Earth-Biogenome-Project-pilot/pull/35/commits/fe7a6895381f8c3bb68fc590009897060c669ef9)

File: modules/local/mitohifi/mitohifi.nf (hunk 5 above)
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/mitohifi.nf]: Done! See this commit: [Moving the default MitoHiFi filenames so they are prefixed with $prefix](https://github.com/NBISweden/Earth-Biogenome-Project-pilot/commit/2c0092815fb66996c063692f645a95c4737cc9c6)

File: modules/local/mitohifi/findmitoref.nf (hunk 3 above)
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/findmitoref.nf]: Does this make sense in combination with the updated `modules.config` file?

File: modules/local/mitohifi/mitohifi.nf (hunk 5 above)
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/mitohifi.nf]: I'll go through everything and check that $prefix is properly used once I'm back

File: modules/local/mitohifi/findmitoref.nf (hunk 3 above)
[the remark this answers is not in the document]
Comment (verku) [reply modules/local/mitohifi/findmitoref.nf]: Thanks for the explanations today! I'll go through all the parameters needed for FindMitoRef and MitoHiFi and check if I implemented them correctly once I'm back.

File: modules/local/mitohifi/mitohifi.nf [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    // There is no conda package available for MitoHiFi
+    // Adding the Docker container from biocontainers found on DockerHub
+    container 'docker://biocontainers/mitohifi:2.2_cv1'
+
+    input:
+    // TODO nf-core: Where applicable all sample-specific information e.g. "id", "single_end", "read_group"
+    //               MUST be provided as an input via a Groovy Map called "meta".
+    //               This information may not be required in some instances e.g. indexing reference genome files:
+    //               https://github.com/nf-core/modules/blob/master/modules/bwa/index/main.nf
+    tuple val(meta), path() // enter the correct variable pointing to the assembled PacBio contigs that should be searched for the mitogenome
```
Comment (verku) [inline modules/local/mitohifi/mitohifi.nf:27]: See assemblies from prepare_input subworkflow

File: nextflow.config [hunk 9]
```diff
@@ -40,6 +40,21 @@ params {
     ncbi_taxonomy      = null
     // MASH
     mash_screen_db     = 'https://gembox.cbcb.umd.edu/mash/refseq.genomes%2Bplasmid.k21s1000.msh'
+    // FINDMITOREF
```
Comment (verku) [inline nextflow.config:43]: Add all parameters that should be set by the user

File: configs/modules.config [hunk 10]
```diff
@@ -230,4 +230,29 @@ process {
     //         saveAs: { filename -> filename.equals('versions.yml') ? null : filename }
     //     ]
     // }
+    // MITOHIFI
+    // withName: 'FINDMITOREF' {
+    //     ext.args   = '--species${meta.species} --email${meta.email} --min_length${meta.min_length}'
+    // TO DO:    not sure what to enter as prefix as this should be the path to the directory that is mounted
+    //     ext.prefix = { ? }
```
Comment (verku) [inline configs/modules.config:237]: Check `meta` from prepare_input subworkflow for sample id

File: modules/local/mitohifi/findmitoref.nf [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+//               https://www.nextflow.io/docs/latest/process.html#ext
+//               where "task.ext" is a string.
+//               Any parameters that need to be evaluated in the context of a particular sample
+//               e.g. single-end/paired-end data MUST also be defined and evaluated appropriately.
+process FINDMITOREF {
+    tag "$meta.id"
+    label 'process_low'
+
+    // There is no conda package available for MitoHiFi
+    // Adding the Docker container from biocontainers found on DockerHub
```
Comment (verku) [inline modules/local/mitohifi/findmitoref.nf:16]: Check FASTK module for message about missing conda package

File: modules/local/mitohifi/mitohifi.nf [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+// TODO nf-core: Optional inputs are not currently supported by Nextflow. However, using an empty
+//               list (`[]`) instead of a file can be used to work around this issue.
+
+process MITOHIFI {
+    tag "$meta.id"
+    label 'process_medium'
+
+    // There is no conda package available for MitoHiFi
+    // Adding the Docker container from biocontainers found on DockerHub
+    container 'docker://biocontainers/mitohifi:2.2_cv1'
```
Comment (verku) [inline modules/local/mitohifi/mitohifi.nf:20]: Check FASTK module for message about missing conda package

File: configs/modules.config [hunk 13]
```diff
@@ -230,4 +230,29 @@ process {
     //         saveAs: { filename -> filename.equals('versions.yml') ? null : filename }
     //     ]
     // }
+    // MITOHIFI
+    // withName: 'FINDMITOREF' {
+    //     ext.args   = '--species${meta.species} --email${meta.email} --min_length${meta.min_length}'
```
Comment (verku) [inline configs/modules.config:235]: Is this correct? Compare to other modules with parameters in `nextflow.config`.

File: configs/modules.config [hunk 14]
```diff
@@ -230,4 +230,29 @@ process {
...
+    //     ext.prefix = { ? }
+    //     publishDir = [
+    //         path: { "$params.outdir/mitohifi/findmitoref" },
+    //         mode: params.publish_mode,
+    //         saveAs: { filename -> filename.equals('versions.yml') ? null : filename }
+    //     ]
+    // }
+    // withName: 'MITOHIFI' {
+    //     ext.args   = '-t 1 -o 5'
+    // TO DO:    not sure what to enter as prefix as it has different values at the moment:
```
Comment (verku) [inline configs/modules.config:246]: This is fixed by commit https://github.com/NBISweden/Earth-Biogenome-Project-pilot/commit/2c0092815fb66996c063692f645a95c4737cc9c6

File: configs/modules.config (hunk 13 above)
Comment (mahesh-panchal) [reply configs/modules.config:235]: Almost. If a variable in task context ( e.g. an input variable like `meta` ) needs to be accessed, then the string should be specified as a closure (anonymous function). If a closure is not used, Nextflow tries to evaluate the string when the config loads before workflow execution. A closure defers evaluation of the string until the task is actually executed.
```suggestion
    //     ext.args   = { '--species${meta.species} --email${meta.email} --min_length${meta.min_length}' }
```

File: configs/modules.config (hunk 14 above)
Comment (verku) [inline configs/modules.config:246]: This needs to be fixed by fixing the input channels and the input files for MitoHiFi (see assemblies from prepare_input and output from FINDMITOREF)

File: modules/local/mitohifi/mitohifi.nf [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    // TODO nf-core: It MUST be possible to pass additional parameters to the tool as a command-line string via the "task.ext.args" directive
+    """
+    mitohifi.py \\
+    -c ${prefix}.fasta \\
```
Comment (verku) [inline modules/local/mitohifi/mitohifi.nf:48]: File names/variables passed to `-c`, `-f` and `-g` need to be updated along with the input channels for this process (`-c` takes assembly from prepare_input, `-f` takes the mitochondrial genome reference in fasta format from findmitoref, and `-g` takes the mitochondrial genome reference in genebank format from findmitoref).

File: modules/local/mitohifi/findmitoref.nf [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    // TODO nf-core: It MUST be possible to pass additional parameters to the tool as a command-line string via the "task.ext.args" directive
+    // TODO        : `--outfolder ${prefix}` might be provided as channel instead (see input)
+    """
+    findMitoReference.py \\
+        --species ${species} \\
+        --email ${email} \\
+        --min_length ${min_length} \\
+        --outfolder ${prefix} \\
```
Comment (verku) [inline modules/local/mitohifi/findmitoref.nf:47]: See to do, provide `--outfolder` as channel

**Comment by verku** [comment]
At the last nf-core hackathon, we worked on a mitohifi module. It is not finished yet but can be implemented here as soon as it is available.

**Comment by verku** [comment]
The nf-core MitoHifi module is now available and can be implemented here.

**Comment by mahesh-panchal** [comment]
Should we close this pull request, or modify this one to add the module? Would you like to add it, or should I?

**Comment by verku** [comment]
I'd prefer to close this PR and make a new one for the nf-core module. I'm quite busy right now, but if you don't have time to add the module, I can keep it on my to do list and work on that when I have some time in a few weeks from now.

**Comment by mahesh-panchal** [comment]
OK. Let's close this PR. The issue is still open, so whoever has time first can take it.

**Comment by verku** [comment]
Sounds good!

**Pull request closed without merging**